lightning/src/blinded_path/payment.rs

```diff
@@ -0,0 +1,285 @@
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub fee_base_msat: u32,
+}
+
+/// Constraints for relaying over a given [`BlindedHop`].
+///
+/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+pub struct PaymentConstraints {
+	/// The maximum total CLTV delta that is acceptable when relaying a payment over this
+	/// [`BlindedHop`].
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub max_cltv_expiry: u32,
+	/// The minimum value, in msat, that may be relayed over this [`BlindedHop`].
+	pub htlc_minimum_msat: u64,
+}
+
+impl Writeable for BlindedPaymentTlvs {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		// TODO: write padding
+		match self {
+			Self::Forward { short_channel_id, payment_relay, payment_constraints, features } => {
+				encode_tlv_stream!(w, {
+					(2, short_channel_id, required),
+					(10, payment_relay, required),
+					(12, payment_constraints, required),
+					(14, features, required),
+				});
+			},
+			Self::Receive { payment_secret, payment_constraints, features } => {
+				encode_tlv_stream!(w, {
+					(6, payment_secret, required),
+					(12, payment_constraints, required),
+					(14, features, required),
+				});
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for BlindedPaymentTlvs {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		_init_and_decode_tlv_stream!(r, {
+			(1, _padding, option),
+			(2, scid, option),
+			(6, payment_secret, option),
+			(10, payment_relay, option),
+			(12, payment_constraints, required),
+			(14, features, required),
+		});
+		if let Some(short_channel_id) = scid {
+			if payment_secret.is_some() { return Err(DecodeError::InvalidValue) }
+			Ok(BlindedPaymentTlvs::Forward {
+				short_channel_id,
+				payment_relay: payment_relay.ok_or(DecodeError::InvalidValue)?,
+				payment_constraints: payment_constraints.0.unwrap(),
+				features: features.0.unwrap(),
+			})
+		} else {
+			if payment_relay.is_some() { return Err(DecodeError::InvalidValue) }
+			Ok(BlindedPaymentTlvs::Receive {
+				payment_secret: payment_secret.ok_or(DecodeError::InvalidValue)?,
+				payment_constraints: payment_constraints.0.unwrap(),
+				features: features.0.unwrap(),
+			})
+		}
+	}
+}
+
+/// Construct blinded payment hops for the given `unblinded_path`.
+pub(super) fn blinded_hops<T: secp256k1::Signing + secp256k1::Verification>(
+	secp_ctx: &Secp256k1<T>, unblinded_path: &[(PublicKey, BlindedPaymentTlvs)], session_priv: &SecretKey
+) -> Result<Vec<BlindedHop>, secp256k1::Error> {
+	let mut blinded_hops = Vec::with_capacity(unblinded_path.len());
+	let mut curr_hop_idx = 0;
+	utils::construct_keys_callback(
+		secp_ctx, unblinded_path.iter().map(|(pk, _)| pk), None, session_priv,
+		|blinded_node_id, _, _, encrypted_payload_ss, _, _| {
+			blinded_hops.push(BlindedHop {
+				blinded_node_id,
+				encrypted_payload: utils::encrypt_payload(&unblinded_path[curr_hop_idx].1, encrypted_payload_ss),
```

Comment:
Hmm, is specifying `rho` clearer? The `rho` functions as an encrypted payload shared secret so I think it's not an inaccurate name per se. Could you clarify what's confusing?

Reply:
Hmm true it's not really inaccurate, but yea I think specifying `rho` could help. When I see ss/shared_secret I usually think of the shared secret formed from ECDH, and then I think of rho/mu/etc. as keys formed from that shared secret, so when I saw `encrypted_payload_ss` it seemed sort of ambiguous whether that was the ECDH shared secret or rho since they're both just 32 byte slices